### ISSUE TITLE
Segmentation fault on sort while refreshing

### ISSUE THREAD

**Issue by mjsir911** [newsboat/newsboat#791]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**:

`newsboat 2.13.0`

Steps to reproduce the issue:

1. open newsboat with a bunch of feeds ( I have like ~60 )

2. while the initial feed list is refreshing, where you see `(x/y) loading ...` on the bottom of the screen:

3. type s (sort) + l (latest first)

**Comment by mjsir911** [comment]
Seems to be fixed in r2.18

**Comment by Minoru** [comment]
Ah, my favourite kind of bugs—the fixed ones! ^_^ Especially since this one looks like a concurrency issue, which are a nightmare to track down.

**Comment by mjsir911** [comment]
Heh, yeah, Really just putting this here for posterity.

I'm still affected by this till Gentoo marks newsboat 2.18 as stable (ironically).

**Comment by Minoru** [comment]
They better hurry, 2.18 will become *stale* (huh, a pun!) in about a month, when 2.19 comes out :)

**Comment by mjsir911** [comment]
Eh, 2.19 has become stable on Gentoo & still dealing with this problem. Will follow up but interested if this can be reproduced

**Comment by mjsir911** [comment]
probably related to #557

**Comment by Minoru** [comment]
I reproduced this on r2.19 tag with:

- a one-line config: `auto-reload yes`
- a urls file of 573 URLs and a corresponding cache file (which weighs a few gigabytes, so pardon me for not sharing it). I didn't check if a smaller set of URLs would trigger this; it's just a nice big set of feeds that takes a while to reload

The steps to reproduce are:

1. start Newsboat, wait for the UI to show up
2. press `gl`
3. if usually crashes after the above, but if it doesn't, type `gf` to switch back to a different sort order, then type `gl` to try crashing it again

These steps did *not* crash on current master (f18894c2c7098553bf37cf474db18829325d9873), so I ran a bisect:

```
$ git bisect start

# Git bisect looks for regressions, but we're looking for the commit that fixed the issue,
# so we have to name the crashing commit "good" and the working one "bad"
$ git bisect bad f18894c2c7098553bf37cf474db18829325d9873
$ git bisect good r2.19

# `sleep 25` waits for Newsboat to start
# `echo -n` avoids a newline after the keys. Without that, Newsboat will enter a feed and then an article.
# `gl` and `gf` switch between different sorts in hope of crashing Newsboat
# The delay after `gl` is larger because it takes longer for that sorting to apply
# `reset` is necessary to fix the mess that ncurses creates because of the segfault
# Exit code of 0 is treated as failure because fixed versions are "bad", crashing are "good"
$ git bisect run sh -c 'make -j9 PROFILE=1 newsboat \
&& (sleep 25 ; \
    for i in `seq 10`; do \
        echo -n gl; \
        sleep 5; \
        echo -n gf; \
        sleep 2; \
    done; \
    echo -n q) \
| ./newsboat --config-file=data/791-config --cache-file=/dev/shm/cache.db; \
result=$?; \
make distclean; \
reset; \
[ $result -eq 0 ] && exit 1 || exit 0'

566c2483b4f37f5a3eaeff70b0688e1862b7f542 is the first bad commit
commit 566c2483b4f37f5a3eaeff70b0688e1862b7f542
Author: Alexander Batischev <eual.jp@gmail.com>
Date:   Tue Sep 1 22:03:20 2020 +0300

    Make FeedContainer::feeds private

 include/feedcontainer.h |  3 +--
 src/controller.cpp      | 14 +++++++-------
 src/opml.cpp            |  2 +-
 src/reloader.cpp        | 12 +++++++-----
 test/feedcontainer.cpp  |  2 +-
 5 files changed, 17 insertions(+), 16 deletions(-)
```

The result makes sense to me: 566c2483b4f37f5a3eaeff70b0688e1862b7f542 is part of #1166 where I fixed a downright stupid flaw: we had two mutexes protecting the same data, with different parts of code using one of those two mutexes, but never both.

That PR was released as part of 2.21, so if you can build from Git, or temporarily unmask 2.21 in Gentoo and test it—please do, @mjsir911. Also good catch regarding #557; I'll go post the same suggestion there as well.

**Comment by Minoru** [comment]
@mjsir911, did you get a chance to test 2.21 and see if it exhibits the bug?

**Comment by Minoru** [comment]
I'm pretty sure this is fixed by 2.21, so I'm closing the issue. Please re-open if you can still reproduce with 2.21.
